model/buildlogger.go

```diff
@@ -88,21 +98,52 @@ func (l *Log) Save() error {
 		l.ID = l.Info.ID()
 	}
 
-	updateResult, err := l.env.GetDB().Collection(buildloggerCollection).ReplaceOne(ctx, bson.M{"_id": l.ID}, l, options.Replace().SetUpsert(true))
+	insertResult, err := l.env.GetDB().Collection(buildloggerCollection).InsertOne(ctx, l)
+	grip.DebugWhen(err == nil, message.Fields{
+		"ns":           model.Namespace{DB: l.env.GetConf().DatabaseName, Collection: buildloggerCollection},
+		"id":           l.ID,
+		"insertResult": insertResult,
+		"op":           "save new buildlogger log",
+	})
+	return errors.Wrapf(err, "problem saving new log %s", l.ID)
+}
+
+// AppendLogChunkInfo adds a new log chunk to the log's chunks array in the
+// database. The environment should not be nil.
+func (l *Log) AppendLogChunkInfo(logChunk LogChunkInfo) error {
+	if l.env == nil {
+		return errors.New("cannot append to a log with a nil environment")
+	}
+	ctx, cancel := l.env.Context()
+	defer cancel()
+
+	if l.ID == "" {
+		l.ID = l.Info.ID()
+	}
+
+	update := bson.M{
+		"$push": bson.M{
+			bsonutil.GetDottedKeyName(logArtifactKey, logArtifactInfoChunksKey): logChunk,
+		},
+	}
+	updateResult, err := l.env.GetDB().Collection(buildloggerCollection).UpdateOne(ctx, bson.M{"_id": l.ID}, update)
 	grip.DebugWhen(err == nil, message.Fields{
 		"ns":           model.Namespace{DB: l.env.GetConf().DatabaseName, Collection: buildloggerCollection},
 		"id":           l.ID,
 		"updateResult": updateResult,
-		"artifact":     l.Artifact,
-		"op":           "save buildlogger log",
+		"logChunkInfo": logChunk,
+		"op":           "append log chunk info to buildlogger log",
 	})
-	return errors.Wrap(err, "problem saving log to collection")
+	if err == nil && updateResult.MatchedCount == 0 {
```

Comment:
I decided to return an error if matched count is 0 (meaning the id DNE in the DB), the alternative is to return matched count and not error. Let me know what you think.

Reply:
sounds good!